File: business/src/main/java/com/radioteria/business/services/misc/PasswordRecoveryCode.java
package com.radioteria.business.services.misc;

import org.springframework.util.DigestUtils;

import java.util.Base64;

public class PasswordRecoveryCode {

    final public static String CODE_DELIMITER = ":";

    private long codeStaleTime;
    private String userEmail;
    private String userDigest;

    public PasswordRecoveryCode(User user, long codeStaleTime) {
        this(user.getEmail(), generateUserDigest(user), codeStaleTime);
    }

    public PasswordRecoveryCode(String userEmail, String userDigest, long codeStaleTime) {
        this.userEmail = userEmail;
        this.userDigest = userDigest;
        this.codeStaleTime = codeStaleTime;
    }

    public long getCodeStaleTime() {
        return codeStaleTime;
    }

    public String getUserEmail() {
        return userEmail;
    }

    public String getUserDigest() {
        return userDigest;
    }

    public String encode() {
        Base64.Encoder encoder = Base64.getEncoder();
        CharSequence[] codeParts = new String[] { userEmail, userDigest, Long.toString(codeStaleTime) };
        return encoder.encodeToString(String.join(CODE_DELIMITER, codeParts).getBytes());
    }

    public static PasswordRecoveryCode decode(String encodedCode) {
        String decodedCode = decodePasswordRecoveryCode(encodedCode);
        String[] codeParts = decodedCode.split(CODE_DELIMITER, 3);

        if (codeParts.length != 3) {
            throw new IllegalArgumentException("Specified code is not valid.");
        }

        String userEmail = codeParts[0];
        String userDigest = codeParts[1];
        long codeStaleTime = Long.parseLong(codeParts[2]);

        return new PasswordRecoveryCode(userEmail, userDigest, codeStaleTime);
    }

    public boolean digestMatches(User user) {
        return userDigest.equals(generateUserDigest(user));
    }

    private static String decodePasswordRecoveryCode(String code) {
        Base64.Decoder decoder = Base64.getDecoder();
        return new String(decoder.decode(code));
    }

    private static String generateUserDigest(User user) {
        String userCredentials = user.getEmail() + user.getPassword();
        return DigestUtils.md5DigestAsHex(userCredentials.getBytes());
    }
}

File: business/src/main/kotlin/com/radioteria/services/user/util/PasswordRecoveryCode.kt
package com.radioteria.services.user.util

import com.radioteria.db.entities.User
import org.springframework.util.DigestUtils

import java.util.Base64

data class PasswordRecoveryCode(val userEmail: String, val userDigest: String, val codeStaleTime: Long) {

    constructor(user: User, codeStaleTime: Long) : this(user.email, generateUserDigest(user), codeStaleTime) {}

    fun encode(): String {
        val encoder = Base64.getEncoder()
        val codeParts = arrayOf(userEmail, userDigest, java.lang.Long.toString(codeStaleTime))
        return encoder.encodeToString(codeParts.joinToString(CODE_DELIMITER).toByteArray())
    }

    fun digestMatches(user: User): Boolean {
        return userDigest == generateUserDigest(user)
    }

    companion object {

        val CODE_DELIMITER = ":"

        fun decode(encodedCode: String): PasswordRecoveryCode {
            val decodedCode = decodePasswordRecoveryCode(encodedCode)
            val codeParts = decodedCode.split(CODE_DELIMITER.toRegex(), 3).toTypedArray()

            if (codeParts.size != 3) {
                throw IllegalArgumentException("Specified code is not valid.")
            }

            val userEmail = codeParts[0]
            val userDigest = codeParts[1]
            val codeStaleTime = java.lang.Long.parseLong(codeParts[2])

            return PasswordRecoveryCode(userEmail, userDigest, codeStaleTime)
        }

        private fun decodePasswordRecoveryCode(code: String): String {
            val decoder = Base64.getDecoder()
            return String(decoder.decode(code))
        }

        private fun generateUserDigest(user: User): String {
            val userCredentials = user.email + user.password
            return DigestUtils.md5DigestAsHex(userCredentials.toByteArray())
        }
    }
}
